### PR TITLE
Prob with shots-noise

### PR DESCRIPTION
**Pull request by multiphaseCFD**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

Add shots-noise to `prob()` in the C++ layer.

**Description of the Change:**

`prob(Observable<StateVectorT> &obs, size_t number_shots)`, `prob(size_t number_shots)` and `prob(const std::vector<size_t> wires, size_t number_shots)` were added for the probabilities of a full system and sub-system. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/core/src/measurements/MeasurementsBase.hpp [hunk 1]
```diff
@@ -291,6 +291,60 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
...
+    /**
+     * @brief Probabilities with shot-noise for a subset of the full system.
+     *
+     * @param num_shots Number of shots.
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+    auto probs(const std::vector<size_t> &wires, const size_t &num_shots)
```
Comment (maliasadi) [inline pennylane_lightning/core/src/measurements/MeasurementsBase.hpp:303]: It can be updated as follows,
```suggestion
    auto probs(const std::vector<size_t> &wires, size_t num_shots)
```
Compilers will optimize unnecessary value copies of arguments for small and simple types.

File: pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp [hunk 2]
```diff
@@ -165,6 +165,32 @@ class Measurements final
...
     }
 
+    /**
+     * @brief Probabilities with shot-noise.
+     *
+     * @param num_shots Number of shots.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+    std::vector<PrecisionT> probs(const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp:175]:
```suggestion
    std::vector<PrecisionT> probs(size_t num_shots) {
```

File: pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp [hunk 3]
```diff
@@ -165,6 +165,32 @@ class Measurements final
...
+     *
+     * @param num_shots Number of shots.
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+
+    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
+                                  const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp:190]:
```suggestion
    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
                                  size_t num_shots) {
```

File: pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPUMPI.hpp [hunk 4]
```diff
@@ -221,6 +221,32 @@ class MeasurementsMPI final
...
+     *
+     * @param num_shots Number of shots.
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+
+    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
+                                  const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPUMPI.hpp:246]:
```suggestion
    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
                                  size_t num_shots) {
```

File: pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp [hunk 5]
```diff
@@ -165,6 +165,32 @@ class Measurements final
...
 
+    /**
+     * @brief Probabilities with shot-noise.
+     *
+     * @param num_shots Number of shots.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+    std::vector<PrecisionT> probs(const size_t &num_shots) {
+        return BaseType::probs(num_shots);
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp:176]: It's really nice to see the base class impl can support the measurement process on all devices even on LGPU :100:

File: pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp [hunk 6]
```diff
@@ -618,6 +618,32 @@ class Measurements final
...
     }
 
+    /**
+     * @brief Probabilities with shot-noise.
+     *
+     * @param num_shots Number of shots.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+    std::vector<PrecisionT> probs(const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp:628]:
```suggestion
    std::vector<PrecisionT> probs(size_t num_shots) {
```

File: pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp [hunk 7]
```diff
@@ -618,6 +618,32 @@ class Measurements final
...
+     *
+     * @param num_shots Number of shots.
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+
+    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
+                                  const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp:643]:
```suggestion
    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
                                  size_t num_shots) {
```

File: pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp [hunk 8]
```diff
@@ -146,6 +146,32 @@ class Measurements final
...
     }
 
+    /**
+     * @brief Probabilities with shot-noise.
+     *
+     * @param num_shots Number of shots.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+    std::vector<PrecisionT> probs(const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:156]:
```suggestion
    std::vector<PrecisionT> probs(size_t num_shots) {
```

File: pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp [hunk 9]
```diff
@@ -146,6 +146,32 @@ class Measurements final
...
+     *
+     * @param num_shots Number of shots.
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     *
+     * @return Floating point std::vector with probabilities.
+     */
+
+    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
+                                  const size_t &num_shots) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:171]:
```suggestion
    std::vector<PrecisionT> probs(const std::vector<size_t> &wires,
                                  size_t num_shots) {
```

**Review by maliasadi — COMMENTED** [review]
Thanks @multiphaseCFD! :100:  Looking forward to see this support in Catalyst :)

**Review by AmintorDusko — APPROVED** [review]
I have just a super small suggestion. Happy to approve!

File: .github/CHANGELOG.md [hunk 10]
```diff
@@ -2,6 +2,9 @@
 
 ### New features since last release
 
+* Add shots-noise support for probs in the C++ layer. Probabilities are calculated from generated samples. All Lightning backends support this support feature. Please note that target wires should be sorted in ascending manner.
```
[the remark this answers is not in the document]
Comment (dime10) [reply .github/CHANGELOG.md:5]: Made a minor edit to your suggestion, as shots in "shot-noise" is typically singular.

File: pennylane_lightning/core/src/measurements/MeasurementsBase.hpp (hunk 1 above)
Comment (dime10) [reply pennylane_lightning/core/src/measurements/MeasurementsBase.hpp:303]: > Compilers will optimize unnecessary value copies of arguments for small and simple types.

Just a side note, but I'm always looking to learn more about C++ and am curious what you mean by this concretely, could you elaborate? :) 

I think I would rather argue that an integer is the smallest unit of data you can copy anyways. Aiming to avoid a copy by using a const reference thus brings no gain, since it incurs a copy (for the pointer) _and_ an additional indirection.

File: pennylane_lightning/core/src/measurements/MeasurementsBase.hpp [hunk 11]
```diff
@@ -291,6 +291,60 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
...
+        size_t num_wires = _statevector.getTotalNumQubits();
+
+        std::vector<PrecisionT> prob_shots(size_t{1} << wires.size(), 0.0);
+
+        for (auto &it : counts_map) {
+            size_t bitVal = 0;
+            for (size_t bit = 0; bit < wires.size(); bit++) {
+                bitVal += ((it.first >> (num_wires - size_t{1} - wires[bit])) &
+                           size_t{1})
+                          << (wires.size() - size_t{1} - bit);
```
Comment (dime10) [inline pennylane_lightning/core/src/measurements/MeasurementsBase.hpp:316]: This formula is a bit hard to follow, might benefit from a comment?

**Review by dime10 — COMMENTED** [review]
Thanks @multiphaseCFD, looks great 💯 The only thing I'm wondering, shouldn't there be a function with an observable and shot argument?

**Comment by multiphaseCFD** [comment]
> Thanks @multiphaseCFD, looks great 💯 The only thing I'm wondering, shouldn't there be a function with an observable and shot argument?

Thanks @dime10 ! `probs(obs, num_shots)` added.

File: pennylane_lightning/core/src/measurements/MeasurementsBase.hpp (hunk 11 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/measurements/MeasurementsBase.hpp:323]: Thanks, a comment added!

File: .github/CHANGELOG.md [hunk 12]
```diff
@@ -2,6 +2,9 @@
 
 ### New features since last release
 
+* Add shot-noise support for probs in the C++ layer. Probabilities are calculated from generated samples. All Lightning backends support this support feature. Please note that target wires should be sorted in ascending manner.
```
Comment (dime10) [inline .github/CHANGELOG.md:5]:
```suggestion
* Add shot-noise support for probs in the C++ layer. Probabilities are calculated from generated samples. All Lightning backends support this feature. Please note that target wires should be sorted in ascending manner.
```

**Review by dime10 — APPROVED** [review]
Looks great 💯 

The only thing that I find a bit confusing is the large amount of variations that exist for different measurement processes, and don't always seem consistent or consistently available. For example, `expval` and `var` provide separate functions for:
- mp(obs)
- mp(obs, shots)

But `probs` combines them into one:
- mp(obs, shots=0)

And that's on top of other variations like `probs()`/`probs(shots)`, and `probs(wires)`/`probs(wires, shots)`, where they are separate as well 😅 

But it's not a blocker :)

File: .github/CHANGELOG.md (hunk 12 above)
Comment (multiphaseCFD) [reply .github/CHANGELOG.md:5]: Sure!

**Comment by multiphaseCFD** [comment]
> Looks great 💯
> 
> The only thing that I find a bit confusing is the large amount of variations that exist for different measurement processes, and that don't always seem consistent or consistently available. For example, `expval` and `var` provide separate functions for:
> 
> * mp(obs)
> * mp(obs, shots)
> 
> But `probs` combines them into one:
> 
> * mp(obs, shots=0)
> 
> And that's on top of other variations like `probs()`/`probs(shots)`, and `probs(wires)`/`probs(wires, shots)`, where they are separate as well 😅
> 
> But it's not a blocker :)

Thanks @dime10 ! The reason why mps except for `probs` accept `(obs)` and `(obs, shots)` is that I don't want to refactor our previous analytic implementations.
